README.md

```diff
@@ -1,39 +1,84 @@
+<LINK TBC>
+
+<To be moved to GH>
+
 ### Migration from v2.1 to v3.0-alpha
 
-Lots of method/type names have changed between v2.0 and v3.0, but there's very little _functionality_ change, so it should be very easy to migrate:
+Lots of method/type names have changed between `v2.1` and `v3.0`, but there's very little _functionality_ change, so it should be very easy to migrate:
 
 * The `[Builder]` attribute is now called `[BuilderFactory]`, to better represent what it is doing.
 * `With(m => m.SubObject)` intended to auto-find any existing builders, is now `WithBuilt(m => m.SubObject)`.
-* `With(m => m.SingleString, "a", "b", "c" )` intended to loop over values, is now `WithSequentialFrom(m => m.SingleString, "a", "b", "c" )`.
-* `WithCollection(...)` is now `WithCreateEnumerableFrom(...)`.
+* `With(m => m.SingleString, "a", "b", "c" )` intended to loop over values, is now `WithOneOf(m => m.SingleString, "a", "b", "c" )`.
+* `WithCollection(...)` is now `WithEnumerable(...)`.
 * The implicit cast from `Builder<T>` to `T` has been removed. Replace it with calling `.Build()` on the builder.
 
+### Migration from v3.0-alpha to v4.0
+
+A Bunch more method/type names have changed between `v3.0` and `v4.0`, but there's very little _functionality_ change, so it should be very easy to migrate:
+
+* The `WithBuilt()` method has been entirely removed, as has the `[BuilderFactory]` attribute that it used.
+  * The `WithBuilt(m => m.SubObject)` method call, intended to identify and use a Registered Builder, is identical to `WithBuilder(m => m.SubObject, TheRegisteredBuilder.New)`. Use the latter wherever you were using the former, explicitly specifying the builder which previously had the `[BuilderFactory]` attribute.
+  * The `WithBuilt(m => m.SubObject)` method call, intended to use the default constructor of the object (i.e. *without* a Builder defined), is now `WithNew(m => m.SubObject)`
+* The `With(m => m.SubObject, someExistingObject)` method call, intended re-use the same `someExistingObject` on every built object, is now `WithSharedRef(m => m.SubObject, someExistingObject)`.
+* `With(m => m.Prop, value)` is now constrained to primitives and other value Types (specifically `where TProp : struct`) to enforce the use of `WithSharedRef` where that's intended.
+* The `WithOneOf(m => m.SingleString, "a", "b", "c" )` method call, intended to loop over values, is now `WithSequentialFrom(m => m.SingleString, "a", "b", "c" )`.
+* `WithEnumerable(...)` is now `WithCreateEnumerableFrom(...)`.
+* `Add(...)` is now `WithAddToCollection(...)`.
+* `WithSetup(...)` is now `WithCustomSetup(...)`.
+* `Build(n)` method is now Eager, returning the result in a `List`, rather than a Lazy `IEnumerable`.
+
+The broad changes to the API were to avoid reduce reliance on Type-based differences in method overload, thereby adding clarity and allowing for more possible behaviours, improving discoverability of the available options.
```

Comment:
Relatedly, the whole readme feels like it's in a little bit of a weird order. If I were writing it, I'd perhaps chose the following order:

* Context / Introduction; Explain the problem it solves (reducing overhead building objects or object dependencies for tests), and how it's intended to be used; Link to migration guides
* Install guide / choice of versions
* Minimal example
* Full list of methods on builder (currently, this seems split between "Notable usages and features" and Setup Methods")
* Full example
* Other notes / help

Reply:
>Also - it feels weird having both v2.1 -> v3.0-alpha and v3.0-alpha -> v4 and also having the full v2.1 -> v4 instructions, in the same document. Maybe having these in a separate migration-guide folder, and link out to them from the core readme?

Agreed. The plan was to pull the alpha-related stuff out of the README entirely, since I strongly suspect that no-one's used it outside of @benbelow.

Dump it in an Issue; close the issue; leave behind a comment for discoverability and have only 2.1 => 4 in the README.

Reply:
Re-ordering suggestion sounds good.

Reply:
Re-ordered.